### PR TITLE
Categorize events as Normal or Warning

### PR DESCRIPTION
**Pull request by aveshagarwal**
This addresses issue #15624 .

### CODE REVIEW

**Comment by aveshagarwal** [comment]
@ncdc @derekwaynecarr @saad-ali @thockin

**Comment by derekwaynecarr** [comment]
I may have missed it, but is there a fieldSelector on severity?

I want to be able to only get events that have a specified severity.

E.g curl /v1/events?fieldSelector=severity=Info

**Comment by aveshagarwal** [comment]
@derekwaynecarr  That makes perfect sense.  But this PR does not include anything related to fieldSelector so I am going to look into it.

**Comment by aveshagarwal** [comment]
Rebased and fixed swagger doc failure.

**Comment by erictune** [comment]
I have reviewed your code.  It looks correct overall.  However, I am unsure of the value of severities in general.  I've commented on #15624

**Comment by aveshagarwal** [comment]
@erictune Thanks for your review. I had used Test for test code in kubernetes that means it should only appear in Testing code not in real deployment of kubernetes cluster. But I can change that for sure.

File: pkg/api/v1/types.go [hunk 1]
```diff
@@ -2194,6 +2194,16 @@ type EventSource struct {
 	Host string `json:"host,omitempty"`
 }
 
+// Valid values for event severity
+const (
+	// Informatian only and will not cause any problems
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: typo: Information

File: pkg/api/v1/types.go [hunk 2]
```diff
@@ -2225,6 +2235,9 @@ type Event struct {
 
 	// The number of times this event has occurred.
 	Count int `json:"count,omitempty"`
+
+	//Severity of this event (Info, Warn, Error)
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: // Severity

File: pkg/api/v1/types.go [hunk 3]
```diff
@@ -2194,6 +2194,16 @@ type EventSource struct {
...
 }
 
+// Valid values for event severity
+const (
+	// Informatian only and will not cause any problems
+	EventSevInfo string = "Info"
+	// These events are to warn that something might go wrong
+	EventSevWarn string = "Warn"
+	// These events are to report that something is high likely to cause problems
+	EventSevError string = "Error"
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: Please don't abbreviate. EventSeverityError

File: pkg/controller/node/nodecontroller.go [hunk 4]
```diff
@@ -758,7 +758,7 @@ func (nc *NodeController) terminatePods(nodeName string, since time.Time) (bool,
 		if remaining < 0 {
 			remaining = 0
 			glog.V(2).Infof("Removing pod %v after %s grace period", pod.Name, grace)
-			nc.recordNodeEvent(nodeName, "TerminatingEvictedPod", fmt.Sprintf("Pod %s has exceeded the grace period for deletion after being evicted from Node %q and is being force killed", pod.Name, nodeName))
+			nc.recordNodeEvent(nodeName, api.EventSevWarn, "TerminatingEvictedPod", fmt.Sprintf("Pod %s has exceeded the grace period for deletion after being evicted from Node %q and is being force killed", pod.Name, nodeName))
```
Comment (bgrant0607) [inline pkg/controller/node/nodecontroller.go]: Info

File: pkg/controller/controller_utils.go [hunk 5]
```diff
@@ -308,11 +308,11 @@ func (r RealPodControl) createPods(nodeName, namespace string, template *api.Pod
 		return fmt.Errorf("unable to create pods, no labels")
 	}
 	if newPod, err := r.KubeClient.Pods(namespace).Create(pod); err != nil {
-		r.Recorder.Eventf(object, "FailedCreate", "Error creating: %v", err)
+		r.Recorder.Eventf(object, api.EventSevError, "FailedCreate", "Error creating: %v", err)
```
Comment (bgrant0607) [inline pkg/controller/controller_utils.go]: Warn

File: pkg/controller/podautoscaler/horizontal.go [hunk 6]
```diff
@@ -98,14 +98,14 @@ func (a *HorizontalController) reconcileAutoscaler(hpa extensions.HorizontalPodA
 
 	scale, err := a.client.Extensions().Scales(hpa.Spec.ScaleRef.Namespace).Get(hpa.Spec.ScaleRef.Kind, hpa.Spec.ScaleRef.Name)
 	if err != nil {
-		a.eventRecorder.Event(&hpa, "FailedGetScale", err.Error())
+		a.eventRecorder.Event(&hpa, api.EventSevError, "FailedGetScale", err.Error())
```
Comment (bgrant0607) [inline pkg/controller/podautoscaler/horizontal.go]: Warn

File: pkg/controller/podautoscaler/horizontal.go [hunk 7]
```diff
@@ -81,7 +81,7 @@ func (a *HorizontalController) computeReplicasForCPUUtilization(hpa extensions.H
 
 	// TODO: what to do on partial errors (like metrics obtained for 75% of pods).
 	if err != nil {
-		a.eventRecorder.Event(&hpa, "FailedGetMetrics", err.Error())
+		a.eventRecorder.Event(&hpa, api.EventSevError, "FailedGetMetrics", err.Error())
```
Comment (bgrant0607) [inline pkg/controller/podautoscaler/horizontal.go]: Warn

File: pkg/controller/podautoscaler/horizontal.go [hunk 8]
```diff
@@ -98,14 +98,14 @@ func (a *HorizontalController) reconcileAutoscaler(hpa extensions.HorizontalPodA
...
-		a.eventRecorder.Event(&hpa, "FailedGetScale", err.Error())
+		a.eventRecorder.Event(&hpa, api.EventSevError, "FailedGetScale", err.Error())
 		return fmt.Errorf("failed to query scale subresource for %s: %v", reference, err)
 	}
 	currentReplicas := scale.Status.Replicas
 
 	desiredReplicas, currentUtilization, err := a.computeReplicasForCPUUtilization(hpa, scale)
 	if err != nil {
-		a.eventRecorder.Event(&hpa, "FailedComputeReplicas", err.Error())
+		a.eventRecorder.Event(&hpa, api.EventSevError, "FailedComputeReplicas", err.Error())
```
Comment (bgrant0607) [inline pkg/controller/podautoscaler/horizontal.go]: Warn

File: pkg/controller/podautoscaler/horizontal.go [hunk 9]
```diff
@@ -146,10 +146,10 @@ func (a *HorizontalController) reconcileAutoscaler(hpa extensions.HorizontalPodA
 		scale.Spec.Replicas = desiredReplicas
 		_, err = a.client.Extensions().Scales(hpa.Namespace).Update(hpa.Spec.ScaleRef.Kind, scale)
 		if err != nil {
-			a.eventRecorder.Eventf(&hpa, "FailedRescale", "New size: %d; error: %v", desiredReplicas, err.Error())
+			a.eventRecorder.Eventf(&hpa, api.EventSevError, "FailedRescale", "New size: %d; error: %v", desiredReplicas, err.Error())
```
Comment (bgrant0607) [inline pkg/controller/podautoscaler/horizontal.go]: Warn

File: pkg/controller/podautoscaler/horizontal.go [hunk 10]
```diff
@@ -168,7 +168,7 @@ func (a *HorizontalController) reconcileAutoscaler(hpa extensions.HorizontalPodA
 
 	_, err = a.client.Extensions().HorizontalPodAutoscalers(hpa.Namespace).UpdateStatus(&hpa)
 	if err != nil {
-		a.eventRecorder.Event(&hpa, "FailedUpdateStatus", err.Error())
+		a.eventRecorder.Event(&hpa, api.EventSevError, "FailedUpdateStatus", err.Error())
```
Comment (bgrant0607) [inline pkg/controller/podautoscaler/horizontal.go]: Warn

File: pkg/controller/service/servicecontroller.go [hunk 11]
```diff
@@ -245,7 +245,7 @@ func (s *ServiceController) processDelta(delta *cache.Delta) (error, bool) {
 				message += " (will not retry): "
 			}
 			message += err.Error()
-			s.eventRecorder.Event(service, "CreatingLoadBalancerFailed", message)
+			s.eventRecorder.Event(service, api.EventSevError, "CreatingLoadBalancerFailed", message)
```
Comment (bgrant0607) [inline pkg/controller/service/servicecontroller.go]: Warn

File: pkg/controller/service/servicecontroller.go [hunk 12]
```diff
@@ -255,14 +255,14 @@ func (s *ServiceController) processDelta(delta *cache.Delta) (error, bool) {
 		cachedService.appliedState = service
 		s.cache.set(namespacedName.String(), cachedService)
 	case cache.Deleted:
-		s.eventRecorder.Event(service, "DeletingLoadBalancer", "Deleting load balancer")
+		s.eventRecorder.Event(service, api.EventSevInfo, "DeletingLoadBalancer", "Deleting load balancer")
 		err := s.balancer.EnsureTCPLoadBalancerDeleted(s.loadBalancerName(service), s.zone.Region)
 		if err != nil {
 			message := "Error deleting load balancer (will retry): " + err.Error()
-			s.eventRecorder.Event(service, "DeletingLoadBalancerFailed", message)
+			s.eventRecorder.Event(service, api.EventSevError, "DeletingLoadBalancerFailed", message)
```
Comment (bgrant0607) [inline pkg/controller/service/servicecontroller.go]: Warn

File: pkg/controller/service/servicecontroller.go [hunk 13]
```diff
@@ -670,7 +670,7 @@ func (s *ServiceController) lockedUpdateLoadBalancerHosts(service *api.Service,
 		return nil
 	}
 
-	s.eventRecorder.Eventf(service, "LoadBalancerUpdateFailed", "Error updating load balancer with new hosts %v: %v", hosts, err)
+	s.eventRecorder.Eventf(service, api.EventSevError, "LoadBalancerUpdateFailed", "Error updating load balancer with new hosts %v: %v", hosts, err)
```
Comment (bgrant0607) [inline pkg/controller/service/servicecontroller.go]: Warn

File: pkg/kubelet/dockertools/manager.go [hunk 14]
```diff
@@ -773,13 +773,13 @@ func (dm *DockerManager) runContainer(
 	dockerContainer, err := dm.client.CreateContainer(dockerOpts)
 	if err != nil {
 		if ref != nil {
-			dm.recorder.Eventf(ref, "Failed", "Failed to create docker container with error: %v", err)
+			dm.recorder.Eventf(ref, api.EventSevError, "Failed", "Failed to create docker container with error: %v", err)
```
Comment (bgrant0607) [inline pkg/kubelet/dockertools/manager.go]: Warn

File: pkg/kubelet/dockertools/manager.go [hunk 15]
```diff
@@ -836,13 +836,13 @@ func (dm *DockerManager) runContainer(
 
 	if err = dm.client.StartContainer(dockerContainer.ID, hc); err != nil {
 		if ref != nil {
-			dm.recorder.Eventf(ref, "Failed",
+			dm.recorder.Eventf(ref, api.EventSevError, "Failed",
```
Comment (bgrant0607) [inline pkg/kubelet/dockertools/manager.go]: Warn

File: pkg/kubelet/kubelet.go [hunk 16]
```diff
@@ -2645,7 +2645,7 @@ func (kl *Kubelet) generatePodStatus(pod *api.Pod) (api.PodStatus, error) {
 	// TODO: Consider include the container information.
 	if kl.pastActiveDeadline(pod) {
 		reason := "DeadlineExceeded"
-		kl.recorder.Eventf(pod, reason, "Pod was active on the node longer than specified deadline")
+		kl.recorder.Eventf(pod, api.EventSevWarn, reason, "Pod was active on the node longer than specified deadline")
```
Comment (bgrant0607) [inline pkg/kubelet/kubelet.go]: Info

File: pkg/kubelet/oom_watcher.go [hunk 17]
```diff
@@ -64,7 +64,7 @@ func (ow *realOOMWatcher) Start(ref *api.ObjectReference) error {
 
 		for event := range eventChannel.GetChannel() {
 			glog.V(2).Infof("Got sys oom event from cadvisor: %v", event)
-			ow.recorder.PastEventf(ref, unversioned.Time{Time: event.Timestamp}, systemOOMEvent, "System OOM encountered")
+			ow.recorder.PastEventf(ref, unversioned.Time{Time: event.Timestamp}, api.EventSevInfo, systemOOMEvent, "System OOM encountered")
```
Comment (bgrant0607) [inline pkg/kubelet/oom_watcher.go]: Warn

File: pkg/kubelet/prober/prober.go [hunk 18]
```diff
@@ -98,13 +98,13 @@ func (pb *prober) probeLiveness(pod *api.Pod, status api.PodStatus, container ap
...
 			if ok {
-				pb.recorder.Eventf(ref, "Unhealthy", "Liveness probe errored: %v", err)
+				pb.recorder.Eventf(ref, api.EventSevError, "Unhealthy", "Liveness probe errored: %v", err)
 			}
 			return probe.Unknown, err
 		} else { // live != probe.Success
 			glog.V(1).Infof("Liveness probe for %q failed (%v): %s", ctrName, live, output)
 			if ok {
-				pb.recorder.Eventf(ref, "Unhealthy", "Liveness probe failed: %s", output)
+				pb.recorder.Eventf(ref, api.EventSevError, "Unhealthy", "Liveness probe failed: %s", output)
```
Comment (bgrant0607) [inline pkg/kubelet/prober/prober.go]: Warn

File: pkg/kubelet/prober/prober.go [hunk 19]
```diff
@@ -133,12 +133,12 @@ func (pb *prober) probeReadiness(pod *api.Pod, status api.PodStatus, container a
...
 			glog.V(1).Infof("readiness probe for %q errored: %v", ctrName, err)
 			if ok {
-				pb.recorder.Eventf(ref, "Unhealthy", "Readiness probe errored: %v", err)
+				pb.recorder.Eventf(ref, api.EventSevError, "Unhealthy", "Readiness probe errored: %v", err)
 			}
 		} else { // ready != probe.Success
 			glog.V(1).Infof("Readiness probe for %q failed (%v): %s", ctrName, ready, output)
 			if ok {
-				pb.recorder.Eventf(ref, "Unhealthy", "Readiness probe failed: %s", output)
+				pb.recorder.Eventf(ref, api.EventSevError, "Unhealthy", "Readiness probe failed: %s", output)
```
Comment (bgrant0607) [inline pkg/kubelet/prober/prober.go]: Warn

**Comment by aveshagarwal** [comment]
@bgrant0607 Updated this PR.

**Comment by ncdc** [comment]
@aveshagarwal rebase please, thanks!

**Comment by aveshagarwal** [comment]
Rebased.

**Comment by erictune** [comment]
LGTM.  There is one item that you might want to change, but I don't feel too strongly about it.

**Comment by erictune** [comment]
Thanks for the fix, and thanks for your patience with my long review latency.

**Comment by aveshagarwal** [comment]
@erictune no problem , and thanks for your review.

**Comment by aveshagarwal** [comment]
@erictune updated this PR.

**Comment by aveshagarwal** [comment]
all checks passed.

**Comment by davidopp** [comment]
@jiangyaoguo

**Comment by aveshagarwal** [comment]
rebased and pushed again.

**Comment by ncdc** [comment]
@k8s-bot unit test this

**Comment by ncdc** [comment]
@ixdy @lavalamp do you have any idea why there aren't any test results coming back for unit/int?

**Comment by ncdc** [comment]
@k8s-bot unit test this

**Comment by ixdy** [comment]
This appears to be failing one of the verification checks:
```
diffing /workspace/kubernetes/docs/api-reference against freshly generated docs
diff -Naupr -I 'Last update' '--exclude=*.md' /workspace/kubernetes/docs/api-reference/extensions/v1beta1/definitions.html /workspace/kubernetes/_tmp/api-reference/extensions/v1beta1/definitions.html
--- /workspace/kubernetes/docs/api-reference/extensions/v1beta1/definitions.html	2015-11-23 17:00:53.532775072 +0000
+++ /workspace/kubernetes/_tmp/api-reference/extensions/v1beta1/definitions.html	2015-11-23 17:03:14.572900895 +0000
@@ -4265,4 +4265,4 @@ Last updated 2015-11-20 03:20:06 UTC
 </div>
 </div>
 </body>
-</html>
+</html>
\ No newline at end of file
diff -Naupr -I 'Last update' '--exclude=*.md' /workspace/kubernetes/docs/api-reference/v1/definitions.html /workspace/kubernetes/_tmp/api-reference/v1/definitions.html
--- /workspace/kubernetes/docs/api-reference/v1/definitions.html	2015-11-23 17:00:53.536775076 +0000
+++ /workspace/kubernetes/_tmp/api-reference/v1/definitions.html	2015-11-23 17:03:07.636894708 +0000
@@ -6898,4 +6898,4 @@ Last updated 2015-11-20 03:19:59 UTC
 </div>
 </div>
 </body>
-</html>
+</html>
\ No newline at end of file
/workspace/kubernetes/docs/api-reference is out of date. Please run hack/update-api-reference-docs.sh
FAILED
```

**Comment by aveshagarwal** [comment]
@ixdy thanks. I will update it soon.

**Comment by aveshagarwal** [comment]
I am debugging why that e2e test (kubectl describe) is failing now as it was not failing before.

**Comment by aveshagarwal** [comment]
I ran the e2e test (kubectl describe) againt local kube cluster and it passed without issues.

```
• [SLOW TEST:35.303 seconds]
Kubectl client
/root/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:910
  Kubectl describe
  /root/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:559
    should check if kubectl describe prints relevant information for rc and pods [Conformance]
    /root/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:558
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Ran 1 of 202 Specs in 35.305 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 201 Skipped PASS
```

**Comment by ncdc** [comment]
@k8s-bot test this

**Comment by erictune** [comment]
LGTM.

File: pkg/api/v1/types.go [hunk 20]
```diff
@@ -2251,6 +2259,9 @@ type Event struct {
 
 	// The number of times this event has occurred.
 	Count int `json:"count,omitempty"`
+
+	// Type of this event (Normal, Warning), new types could be added in the future
+	Type string `json:"type,omitempty"`
```
Comment (lavalamp) [inline pkg/api/v1/types.go]: Can we change this name to "severity"?

File: pkg/client/record/event.go [hunk 21]
```diff
@@ -60,13 +61,13 @@ type EventRecorder interface {
 	// 'message' is intended to be human readable.
 	//
 	// The resulting event will be created in the same namespace as the reference object.
-	Event(object runtime.Object, reason, message string)
+	Event(object runtime.Object, eventtype, reason, message string)
```
Comment (lavalamp) [inline pkg/client/record/event.go]: eventType, but even better as eventSeverity, or best of all, just "severity"

File: pkg/api/v1/types.go (hunk 20 above)
Comment (lavalamp) [inline pkg/api/v1/types.go]: I guess this name was @bgrant0607's idea-- I still think "type" is an ambiguous name.

File: pkg/client/record/event.go (hunk 21 above)
Comment (lavalamp) [inline pkg/client/record/event.go]: If we keep "type" as a name, this should just be "type", not eventtype.

Comment (bgrant0607) [inline pkg/client/record/event.go]: Yes, I intentionally wanted to avoid the term "severity".

File: pkg/api/v1/types.go (hunk 20 above)
Comment (bgrant0607) [inline pkg/api/v1/types.go]: It's intentionally ambiguous. I don't want this to evolve in the direct of log severity levels, which are very hard to use in my experience. It's just a coarse bucketing of Reasons along some useful dimensions, TBD. Like Reason, it shouldn't be considered to be an enum.

The Real Solution to filtering is something like separately defined alerting rules.

Comment (lavalamp) [inline pkg/api/v1/types.go]: I see your point about severity levels. I agree external filtering/alerting is the way to go, and is in fact what I wanted events to go into all along, but no time :(

I tend to think that it'd be better without type at all, considering the above. I don't see how local determinations of normal vs warn can be reliably made.

**Pull request merged**